pkg/models/iam/im/im.go

```diff
@@ -170,6 +206,108 @@ func (im *defaultIMOperator) CreateUser(user *iamv1alpha2.User) (*iamv1alpha2.Us
 	return user, nil
 }
 
+func (im *defaultIMOperator) authRateLimitEnabled() bool {
+	if im.authenticateRateLimiterMaxTries <= 0 || im.authenticateRateLimiterDuration == 0 {
+		return false
+	}
+	return true
+}
+
+func (im *defaultIMOperator) authRateLimitExceeded(user *iamv1alpha2.User) bool {
+	if !im.authRateLimitEnabled() {
+		return false
+	}
+	count := 0
+	for _, condition := range user.Status.Conditions {
+		if condition.Status == iamv1alpha2.ConditionTrue &&
+			condition.Type == iamv1alpha2.LoginFailure &&
+			time.Now().Sub(condition.LastTransitionTime.Time) <= im.authenticateRateLimiterDuration {
+			count++
+		}
+	}
+	return count >= im.authenticateRateLimiterMaxTries
+}
+
+func (im *defaultIMOperator) authFailRecord(user *iamv1alpha2.User, err error) {
+	if !im.authRateLimitEnabled() {
+		return
+	}
+	user = user.DeepCopy()
+	condition := iamv1alpha2.UserCondition{
+		Type:               iamv1alpha2.LoginFailure,
+		Status:             iamv1alpha2.ConditionTrue,
+		LastTransitionTime: metav1.Time{Time: time.Now()},
+		Reason:             "Authentication Failed",
+		Message:            err.Error(),
+	}
+	if user.Status.Conditions == nil {
+		user.Status.Conditions = make([]iamv1alpha2.UserCondition, 0)
+	}
+	user.Status.Conditions = append(user.Status.Conditions, condition)
+	conditions := make([]iamv1alpha2.UserCondition, 0)
+	// remove out date and out range record
+	record := 0
+	for i := len(user.Status.Conditions); i > 0; i-- {
+		condition := user.Status.Conditions[i-1]
+		if condition.Status == iamv1alpha2.ConditionTrue &&
+			condition.Type == iamv1alpha2.LoginFailure {
+			if record < im.authenticateRateLimiterMaxTries ||
+				time.Now().Sub(condition.LastTransitionTime.Time) < im.authenticateRateLimiterDuration {
+				conditions = append([]iamv1alpha2.UserCondition{condition}, conditions...)
+				record++
+			}
+		} else {
+			conditions = append([]iamv1alpha2.UserCondition{condition}, conditions...)
+		}
+	}
+	user.Status.Conditions = conditions
+	_, err = im.ksClient.IamV1alpha2().Users().Update(user)
+	if err != nil {
+		klog.Warning(err)
+	}
+}
+
+func (im *defaultIMOperator) AuthSuccessRecord(username string, token string) {
```

Comment:
same here, for example, create a Login object for every login attempt, count consecutive failed login attempts
```
kind: Login
metadata:
  name: admin-xxxx
  createTimestamp: 2020-07-01T14:00:00
  labels:
    user: admin
  ownerReference:
     kind: User
spec:
  sourceIP: xxx.xxx.xxx.xxx
  type: Succeess
  
```